### PR TITLE
More flexible widget lifetimes

### PR DESCRIPTION
**Pull request by nmandrus1**
# Add separate render lifetime parameter to improve widget flexibility

## Motivation
The current implementation of `ListBuilder` requires widgets to have a `'static` lifetime due to how the closure is typed. This constraint is unnecessarily restrictive and prevents the use of widgets that contain references with non-static lifetimes. By separating out the render lifetime from the closure's lifetime, we can support a broader range of widget types while maintaining safety.

## Technical Changes
- Added a new lifetime parameter `'render` to `ListBuilder<T>` to become `ListBuilder<'render, T>`
- Modified the `ListBuilderClosure` type alias to include the `'render` lifetime bound
- Updated `ListView` to carry both a `'block` lifetime (for the optional Block) and the new `'render` lifetime
- Added a `PhantomData<&'render T>` marker to `ListBuilder` to properly track the lifetime parameter
- Updated all implementations and trait bounds throughout the codebase to accommodate the new lifetime parameter
- Modified example code to use the new lifetime parameters

## Benefits
- Widgets can now contain references with lifetimes shorter than `'static`
- More flexible widget composition while maintaining Rust's lifetime safety guarantees
- No runtime performance impact - changes are purely at the type system level
- All existing functionality and tests continue to pass

## Examples
Before this change, widgets containing references had to ensure those references were `'static`. Now they can be bound to any appropriate lifetime:

```rust
// Before: Required 'static
let builder = ListBuilder::new(|ctx| make_widget_with_static_refs());

// After: Can use non-static references
let text = String::from("example");
let builder = ListBuilder::new(|ctx| make_widget_with_refs(&text));
```

Here is the code that inspired this change: 
```rust
struct NotebookCell<'a> {
    cell: &'a Cell,
    style: Style,
}

impl<'a> NotebookCell<'a> {
    fn new(cell: &'a Cell) -> Self {
        Self {
            cell,
            style: Style::default(),
        }
    }
}

impl<'a> Widget for NotebookCell<'a> {
    fn render(self, area: Rect, buf: &mut Buffer) {
        // Custom rendering with borders
        let inner = {
            let block = Block::default()
                .borders(Borders::ALL)
                .border_style(self.style);
            block.clone().render(area, buf);
            block.inner(area)
        };

        // Render cell content
        let content = self.cell.source().join("\n");
        // Later: Add syntax highlighting here
        Line::from(content).render(inner, buf);
    }
}

struct NotebookWidget<'render> {
    notebook: &'render Notebook,
}

impl<'render> NotebookWidget<'render> {
    fn new(notebook: &'render Notebook) -> Self {
        Self { notebook }
    }
}

impl<'render> StatefulWidget for NotebookWidget<'render> {
    type State = ListState;

    fn render(self, area: Rect, buf: &mut Buffer, state: &mut Self::State) {
        let item_count = self.notebook.cells.len();

        let builder = ListBuilder::new(move |context| {
            let cell = &self.notebook.cells[context.index];
            let mut widget = NotebookCell::new(cell);

            if context.is_selected {
                widget.style = Style::default().bg(BLUE.c900);
            }

            // Calculate height based on content
            let height = cell.source().len() as u16 + 2;

            (widget, height)
        });

        ListView::new(builder, item_count)
            .block(
                Block::default()
                    .borders(Borders::ALL)
                    .border_type(BorderType::Rounded)
                    .title("Notebook Cells"),
            )
            .render(area, buf, state);
    }
}
```

This is code for a TUI front-end for a jupyter notebook. I wanted the widget-list to render references to cells, and since the `NotebookCell` did not have a static lifetime this code didn't compile. 

The big change even get's elided by the compiler: 

```rust
        let builder: ListBuilder<'render, NotebookCell<'render>> = ListBuilder::new(move |context| {
        ... 
```

## Merging

I believe this idea is on the right track, but some more thorough examination of the example code, and perhaps even a new example, is in order. I also believe that in some cases this change is overkill and having to juggle several lifetime parameters may be unnecessary. Let me know your thoughts!

### CODE REVIEW

**Comment by preiter93** [comment]
Thx a lot @nmandrus1 for the PR and especially the detailed description! 

I think you are right, the ListBuilder is too restrictive. I've usually just cloned the data, but if we relax the lifetime, we can avoid these clones.

Let me think about it and I'll give it a closer look on the weekend. I want to check if this change will spill out to the user or if the lifetime will be elided in most cases. Depending on that we need to bump the version or not (theoretically its a breaking change, but practically I don't know yet).

I've condensed your example a bit further:

```rust
struct Notebook {}

struct NotebookWidget<'render> {
    notebook: &'render Notebook,
}

struct NotebookCell {}

impl NotebookCell {
    fn new<'a>(_: &'a Notebook) -> Self {
        Self {}
    }
}

impl Widget for NotebookCell {
    fn render(self, _: Rect, _: &mut Buffer) {}
}

impl<'render> StatefulWidget for NotebookWidget<'render> {
    type State = ListState;

    fn render(self, area: Rect, buf: &mut Buffer, state: &mut Self::State) {
        let builder = ListBuilder::new(move |_| {
            let notebook = &self.notebook;
            let widget = NotebookCell::new(notebook);

            (widget, 0)
        });

        ListView::new(builder, 0).render(area, buf, state);
    }
}
```

File: src/view.rs [hunk 1]
```diff
@@ -136,21 +138,23 @@ pub struct ListBuildContext {
...
 /// A type alias for the closure.
-type ListBuilderClosure<T> = dyn Fn(&ListBuildContext) -> (T, u16);
+type ListBuilderClosure<'render, T> = dyn Fn(&ListBuildContext) -> (T, u16) + 'render;
 
 /// The builder to for constructing list elements in a `ListView<T>`
-pub struct ListBuilder<T> {
-    closure: Box<ListBuilderClosure<T>>,
+pub struct ListBuilder<'render, T> {
+    closure: Box<ListBuilderClosure<'render, T>>,
+    _phantom: PhantomData<&'render T>,
```
Comment (preiter93) [inline src/view.rs:146]: do we need this? We already use the lifetime in the closure.

File: src/view.rs [hunk 2]
```diff
@@ -10,12 +12,12 @@ use crate::{utils::layout_on_viewport, ListState};
 /// A struct representing a list view.
 /// The widget displays a scrollable list of items.
 #[allow(clippy::module_name_repetitions)]
-pub struct ListView<'a, T> {
+pub struct ListView<'block, 'render, T> {
```
Comment (preiter93) [inline src/view.rs:15]: Can we get by with just one lifetime here? 

This way we don't have to juggle 2 lifetimes. And all examples will compile without changes which is a good indication that most users don't need changes.

File: src/view.rs (hunk 1 above)
Comment (nmandrus1) [reply src/view.rs:146]: I needed it at some point when I was getting things working, but I removed it and all the tests still pass! It seem's totally unnecessary now.

File: src/view.rs (hunk 2 above)
Comment (nmandrus1) [reply src/view.rs]: Yes we totally can, and that makes way more sense anyways since there can't be many cases where these need to be separate lifetimes.

**Comment by nmandrus1** [comment]
Also I accidentally included a dummy file "hx" which doesn't do anything. Working on removing that.

**Comment by preiter93** [comment]
I've tested the change against various dependents of this library and none of them needed adjustment. 
So I decided to publish a new version without major version change as v0.13.1.

Thx again @nmandrus1 for the contribution, really appreciate it.

**Pull request merged**